crates/wascc-host/src/host_controller.rs

```diff
@@ -254,20 +270,53 @@ impl Handler<StartProvider> for HostController {
     type Result = ResponseActFuture<Self, Result<()>>;
 
     fn handle(&mut self, msg: StartProvider, ctx: &mut Context<Self>) -> Self::Result {
+        let sub = msg.provider.claims.subject.to_string();
+        if self.providers.contains_key(&sub) {
+            error!("Aborting attempt to start already running provider {}", sub);
+            return Box::pin(
+                async move { Err(format!("Cannot start already running provider {}", sub).into()) }
+                    .into_actor(self),
+            );
+        }
+
+        trace!(
+            "Starting provider {} per request",
+            msg.provider.claims.subject
+        );
+
         let seed = self.kp.as_ref().unwrap().seed().unwrap();
         let s = seed.clone();
         let mw = self.mw_chain.clone();
 
         let provider = msg.provider;
-        let sub = provider.claims.subject.to_string();
+
         let capid = provider.claims.metadata.as_ref().unwrap().capid.to_string();
         let binding_name = provider.binding_name.to_string();
         let provider_id = provider.claims.subject.to_string();
         let image_ref = msg.image_ref.clone();
         let ir2 = msg.image_ref.clone();
 
+        if let Err(e) = NativeCapabilityHost::try_new(
+            provider.clone(),
+            mw.clone(),
+            KeyPair::from_seed(&seed).unwrap(),
+            image_ref.clone(),
+        ) {
+            error!("Failed to create a native capability provider host: {}", e);
+            return Box::pin(
+                async move {
+                    Err(format!("Failed to create native capability provider host: {}", e).into())
+                }
+                .into_actor(self),
+            );
+        } // dynamically loaded library is dropped here
+
         let new_provider = SyncArbiter::start(1, move || {
-            //TODO: get rid of this unwrap - it can take out the entire host controller
+            // I know we're calling try_new here twice. Problem is I need to detect the failure
+            // and reject the start attempt (see above)... and I can't instantiate the actor outside this
+            // closure because it's not cloneable. We could in theory fix this by making
+            // the NCH cloneable which means we could instantiate once then return the clone
+            // in this closure... doesn't feel worth the effort (yet).
```

Comment:
Denial ✔️ 
Anger ✔️ 
Bargaining ◀️

Reply:
This may be the single best PR comment I have ever received.